clippy_lints/src/returns.rs

```diff
@@ -176,12 +176,37 @@ fn stmt_needs_never_type(cx: &LateContext<'_>, stmt_hir_id: HirId) -> bool {
         })
 }
 
+///
+/// The expression of the desugared `try` operator is a match over an expression with type:
+/// `ControlFlow<A:Result<Infallible, E>, B:Result<_, E'>>`, with final type `B`.
+/// If E and E' are the same type, then there is no error conversion happening.
+/// Error conversion happens when E can be transformed into E' via a `From` or `Into` conversion.
```

Comment:
@Jarcho You usually know about utils in rustc/Clippy. Do you happen to know of a better way to do this, that doesn't require matching on the exact desugaring of the `?` operator?

Reply:
If not, this LGTM and you can r=me

Reply:
Isn't the lint suggesting to remove the `return` but leave the `?`? There isn't a need to check for type changes in `FromResidual` for that. There is an error with the lint where it doesn't verify the existence of the `Err` constructor before linting.

Reply:
Oh right. The `?` is a red herring here... The suggestion of removing the `return` in the given test/example compiles just fine. It only gives an additional warning that the result is unused. But that is because the `map` turns the `Result<()>` returned by `bar` into a `Result<Result<()>>`, which is a bit nonsensical. 

With the given examples and reproducers, I would call this a non-issue and that the lint does the right thing here. I think I need a more realistic example to understand what the actual issue is in this case. The lint definitely doesn't suggest to remove the `?`.

Reply:
The better question is: Why does it suggest to remove the `return` here in the first place: The lint [documentation](clippy::needless_return_with_question_mark) only talks about the explicit case of `return Err(...)` and not more complex expressions, like in the example. Either the documentation is outdated, or the lint is misbehaving a lot.

Reply:
Could it be that the desugaring is actually causing the misbehaviour here?

Reply:
I'm not sure. I'm not familiar with the code of this lint, sadly :/

Reply:
It's been a while since I looked into this to be honest and I have sort of forgot. I will do a deep dive tonight or tomorrow again and add proper context.

Reply:
I initially thought the issue was related to branching checks, but it turns out that if there's no need for conversion the lint wasn't triggering.

That lead me to investigating the desugaring to figure out what was up.